### PR TITLE
Monitor training

### PR DESCRIPTION
**Pull request by mmaaz60**
Display remaining training time on the training chart.

### CODE REVIEW

**Comment by AlexeyAB** [comment]
@mmaaz60 Hi, Thanks!

1. Can you solve conflicts?
2. Also better to use rolling average with `alpha=0.01`, because it much more simpler and requires fewer changes.

Then I will merge this PR.

**Comment by mmaaz60** [comment]
> @mmaaz60 Hi, Thanks!
> 
> 1. Can you solve conflicts?
> 2. Also better to use rolling average with `alpha=0.01`, because it much more simpler and requires fewer changes.
> 
> Then I will merge this PR.

Thank You @AlexeyAB,

I have implemented rolling average with `alpha = 0.01` and resolved the conflicts as well. Please have a look at this and let me know if you want to implement further changes.

**Stay Safe**,
Maaz

**Comment by AlexeyAB** [comment]
@mmaaz60 Thanks.

> ./src/detector.c:349:45: error: 'i' undeclared (first use in this function)
>          time_remaining = (net.max_batches - i)*(end - start) / 60 / 60;

Use `iteration` instead of `i` for the current iterations number.

----

Can we use ?
`time_remaining = (net.max_batches - i)*(what_time_is_it_now() - time) / 60 / 60;`
instead of
`time_remaining = (net.max_batches - i)*(end - start) / 60 / 60;`

**Comment by mmaaz60** [comment]
> @mmaaz60 Thanks.
> 
> > ./src/detector.c:349:45: error: 'i' undeclared (first use in this function)
> > time_remaining = (net.max_batches - i)*(end - start) / 60 / 60;
> 
> Use `iteration` instead of `i` for the current iterations number.

Thank You @AlexeyAB for the correction. I did not compile the repository after resolving the conflicts. I have updated the code to use `iteration` for current iteration number.

> Can we use ?
> `time_remaining = (net.max_batches - i)*(what_time_is_it_now() - time) / 60 / 60;`
> instead of
> `time_remaining = (net.max_batches - i)*(end - start) / 60 / 60;`

We can use this. I think it will be more precise if we add `load_time` into this as well. For example 
```
time_remaining = (net.max_batches - iteration)*(what_time_is_it_now() - time + load_time) / 60 / 60;
```
This way we will get relatively better estimate of remaining training time and ensure minimum code addition as well. Please have a look at this and let me know if any other change is required.

**Stay Safe**,
Maaz

File: src/detector.c [hunk 1]
```diff
@@ -341,7 +344,15 @@ void train_detector(char *datacfg, char *cfgfile, char *weightfile, int *gpus, i
...
-        draw_train_loss(windows_name, img, img_size, avg_loss, max_img_loss, iteration, net.max_batches, mean_average_precision, draw_precision, "mAP%", dont_show, mjpeg_port);
+        time_remaining = (net.max_batches - iteration)*(what_time_is_it_now() - time + load_time) / 60 / 60;
+        if (iteration > 1){  // ignore the first iteration
+            if (iteration == 2){
+                avg_t_minus_1 = time_remaining;
+            }
+            avg_t = alpha * time_remaining + (1 - alpha) * avg_t_minus_1;
+            avg_t_minus_1 = avg_t;
+        }
+        draw_train_loss(windows_name, img, img_size, avg_loss, max_img_loss, iteration, net.max_batches, mean_average_precision, draw_precision, "mAP%", dont_show, mjpeg_port, avg_t);
```
Comment (AlexeyAB) [inline src/detector.c]:
```suggestion
        time_remaining = (net.max_batches - iteration)*(what_time_is_it_now() - time + load_time) / 60 / 60;
         // set initial value, even if we resume training from 10 000 iteration
        if (avg_time < 0) avg_time = time_remaining;
        else avg_time = alpha_time * time_remaining + (1 - alpha_time) * avg_time;
#ifdef OPENCV
        draw_train_loss(windows_name, img, img_size, avg_loss, max_img_loss, iteration, net.max_batches, mean_average_precision, draw_precision, "mAP%", dont_show, mjpeg_port, avg_time);
```

File: src/detector.c [hunk 2]
```diff
@@ -167,7 +167,10 @@ void train_detector(char *datacfg, char *cfgfile, char *weightfile, int *gpus, i
     //printf(" imgs = %d \n", imgs);
 
     pthread_t load_thread = load_data(args);
+
     int count = 0;
+    double time_remaining, avg_t_minus_1, avg_t, alpha = 0.01;
```
Comment (AlexeyAB) [inline src/detector.c]:
```suggestion
    double time_remaining, avg_time = -1, alpha_time = 0.01;
```

File: src/detector.c [hunk 3]
```diff
@@ -341,7 +344,15 @@ void train_detector(char *datacfg, char *cfgfile, char *weightfile, int *gpus, i
             draw_precision = 1;
         }
 #ifdef OPENCV
```
Comment (AlexeyAB) [inline src/detector.c]:
```suggestion

```

**Comment by mmaaz60** [comment]
> 1. let use `avg_time` instead of `avg_t` because postfix `_t` usually is used for types
> 2. use `if (avg_time < 0){` in both Classifier and Detector, so it will correctly even you resume training from iteration=10000
> 3. Also add output for `time_remaining`   there, it should be shown even if Darknet compiled without OpenCV https://github.com/AlexeyAB/darknet/blob/570f595247db2373e0ee23fe0d1af76697fd88df/src/detector.c#L295



> 1. let use `avg_time` instead of `avg_t` because postfix `_t` usually is used for types
> 2. use `if (avg_time < 0){` in both Classifier and Detector, so it will correctly even you resume training from iteration=10000
> 3. Also add output for `time_remaining`   there, it should be shown even if Darknet compiled without OpenCV https://github.com/AlexeyAB/darknet/blob/570f595247db2373e0ee23fe0d1af76697fd88df/src/detector.c#L295

Thanks @AlexeyAB 

I have implemented the changes as suggested by you. Some tests are failing on macOS as per `continuous-integration/travis-ci/pr`. Couldn't figure out the reason. Please have a look.

**Stay Safe,**
Maaz

**Pull request merged**